pkg/retry/unconfirmed_message_handler.go

```diff
@@ -2,120 +2,78 @@ package retry
-	defaultBaseInterval = 1 * time.Minute
+	log = logging.LoggerForModule()
 )
 
 // UnconfirmedMessageHandlerImpl informs the caller whether a resending should happen based on receiving ACK messsages.
 // Assumption: Time to receive an ACK is generally much shorter than the interval between sending consequtive messages.
 type UnconfirmedMessageHandlerImpl struct {
-	// baseInterval defines the delay after which we resend a message
-	baseInterval time.Duration
 	// resendInterval is a multiply of the baseInterval and defines how much time should pass before the next retry
 	resendInterval time.Duration
-	// ticker controls the delay after which we resend a message
-	ticker *time.Ticker
-	// ch will produce a message when the message should be resent
-	ch chan struct{}
-	// retry counts the number of retries for a given message
-	retry int
-	// retryMux is a mutex for retry variable
-	retryMux *sync.Mutex
-	// numUnackedSendings counts how many sendings occured (not retries) since the last ack
-	numUnackedSendings int
 	// ctx is a context that can be used to stop this object
 	ctx context.Context
+
+	b         *backoff.ExponentialBackOff
+	resendFun func() error
+	gotAck    concurrency.Signal
 }
 
 // NewUnconfirmedMessageHandler returns a new running UnconfirmedMessageHandlerImpl.
 // It can be stopped by canceling the context.
-func NewUnconfirmedMessageHandler(ctx context.Context, resendInterval time.Duration) *UnconfirmedMessageHandlerImpl {
+func NewUnconfirmedMessageHandler(ctx context.Context, b *backoff.ExponentialBackOff) *UnconfirmedMessageHandlerImpl {
 	nsr := &UnconfirmedMessageHandlerImpl{
-		baseInterval:   resendInterval,
-		resendInterval: resendInterval,
-		ticker:         time.NewTicker(resendInterval),
-		ch:             make(chan struct{}),
-		retry:          0,
-		retryMux:       &sync.Mutex{},
-		ctx:            ctx,
+		ctx: ctx,
+		b:   b,
+		resendFun: func() error {
+			return nil
+		},
+		gotAck: concurrency.NewSignal(),
 	}
-	nsr.ticker.Stop()
-	nsr.run()
+	nsr.b.Reset()
 	return nsr
 }
 
-// RetryCommand returns a channel that will produce a message when sending should be retried
-func (s *UnconfirmedMessageHandlerImpl) RetryCommand() <-chan struct{} {
-	return s.ch
-}
-
-// run starts the ticker, so that resending can happen after defined time without ACK passes
-func (s *UnconfirmedMessageHandlerImpl) run() {
-	go func() {
-		defer close(s.ch)
-		for {
-			select {
-			case <-s.ticker.C:
-				s.retryLater()
-				log.Infof("Suggesting to resend, retry %d (next retry in %s)", s.retry, s.resendInterval)
-				s.ch <- struct{}{}
-			case <-s.ctx.Done():
-				return
-			}
+// SetOperation sets function that should produce a message
+func (s *UnconfirmedMessageHandlerImpl) SetOperation(send func() error) {
+	s.resendFun = func() error {
+		go func() {
+			_ = send()
+		}()
+		select {
+		case <-s.gotAck.Done():
+			log.Infof("Got ack!")
+			s.gotAck.Reset()
+			return nil
+		case <-time.After(s.b.NextBackOff()):
```

Comment:
If the Nack is expected to arrive after `s.b.NextBackOff()` time, can't we run into problems by checking if exactly that amount of time has passed before we received one? I would assume that we need some slack here to account for latency

Reply:
The question would be what is a goo latency? :) I went here with 0s extra latency, cause the ACK arriving 1ms later would not cause any extra waiting (another case of that switch would be triggered).